File: src/main/java/no/nav/jfr/generell/infrastructure/metrics/Metrics.java
package no.nav.jfr.generell.infrastructure.metrics;

import io.prometheus.client.Counter;
import io.prometheus.client.Gauge;
import no.nav.jfr.generell.infrastructure.kafka.EnrichedKafkaEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;

public class Metrics {
    private static final Logger log = LoggerFactory.getLogger(Metrics.class);

    private static final String MANUELL = "Manuell journalforing";
    private static final String AUTO = "Automatisk journalforing";

    private static final Counter jfrProcessCounter = Counter.build(
            "jfr_generell_process_counter",
            "Teller antall journalposter som behandles av jfr-generell. Setter på labels på om det burde blitt behandlet maskinelt/manuelt og om den blir behandlet maskinelt/manuelt")
            .labelNames("result", "desired", "tema", "kanal", "skjema")
            .register();

    private static final Counter feilregOppgaverCounter = Counter.build(
            "jfr_generell_feilregistrerte_oppgaver_counter",
            "Teller antall ganger vi må avbryte autoatiske prosess på en slik måte at en oppgave må feilregistreres")
            .labelNames("ableToFeilreg", "tema", "kanal", "skjema")
            .register();

    private static final Counter functionalReqFailCounter = Counter.build(
            "jfr_generell_functional_req_fail_counter",
            "Teller funksjonelle hindringer for automatisk journalforing")
            .labelNames("reason", "tema", "kanal", "skjema")
            .register();

    private static final Counter invalidJournalpostStatusCounter = Counter.build(
            "jfr_generell_invalid_journalpost_status_counter",
            "Teller journalposter fra SAF med ugyldig status")
            .labelNames("status", "tema", "kanal", "skjema")
            .register();

    private static final Counter retryCounter = Counter.build(
            "jfr_generell_retry_counter",
            "Teller antall journalposter hver gang vi vil prøve på nytt senere")
            .labelNames("transformer", "numErrors", "tema", "kanal", "skjema")
            .register();

    private static final Gauge retrystoreGauge = Gauge.build(
            "jfr_generell_retry_gauge",
            "Viser hvor mange JP som ligger på retrystore")
            .labelNames("transformer")
            .register();

    public static void incJfrManuallProcess(EnrichedKafkaEvent enrichedKafkaEvent, boolean desiredAutomaticJfr){
        String desired = "";
        String skjema = "null";
        try {
            MDC.put("CORRELATION_ID", enrichedKafkaEvent.getCorrelationId());
            desired = (desiredAutomaticJfr ? AUTO : MANUELL);
            if(enrichedKafkaEvent.getJournalpost() != null ){
                skjema = enrichedKafkaEvent.getSkjema();
            }
            jfrProcessCounter.labels(MANUELL, desired, enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getKafkaEvent().getMottaksKanal(), String.valueOf(skjema)).inc();
        }catch (Exception e){
            log.error("Klarte ikke å inkrementere metrikk jfrProcessCounter med verdier: '{}' '{}' '{}' '{}' '{}'", MANUELL, desired, enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getKafkaEvent().getMottaksKanal(), skjema, e);
        }
    }

    public static void incJfrAutoProcess(EnrichedKafkaEvent enrichedKafkaEvent){
        try {
            MDC.put("CORRELATION_ID", enrichedKafkaEvent.getCorrelationId());
            jfrProcessCounter.labels(AUTO, AUTO, enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getKafkaEvent().getMottaksKanal(), String.valueOf(enrichedKafkaEvent.getSkjema())).inc();
        }catch (Exception e){
            log.error("Klarte ikke å inkrementere metrikk jfrProcessCounter med verdier: '{}' '{}' '{}' '{}' '{}'", MANUELL, AUTO, enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getKafkaEvent().getMottaksKanal(), String.valueOf(enrichedKafkaEvent.getSkjema()), e);
        }
    }

    public static void incFeilregCounter(EnrichedKafkaEvent enrichedKafkaEvent, boolean sucessfullFeilreg){
        try {
            MDC.put("CORRELATION_ID", enrichedKafkaEvent.getCorrelationId());
            feilregOppgaverCounter.labels(Boolean.toString(sucessfullFeilreg), enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getKafkaEvent().getMottaksKanal(), String.valueOf(enrichedKafkaEvent.getSkjema())).inc();
        }catch (Exception e){
            log.error("Klarte ikke å inkrementere metrikk feilregOppgaverCounter med verdier: '{}' '{}' '{}' '{}'", sucessfullFeilreg, enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getKafkaEvent().getMottaksKanal(), enrichedKafkaEvent.getSkjema(), e);
        }
    }

    public static void incFailFunctionalRequirements(String reason, EnrichedKafkaEvent enrichedKafkaEvent){
        try {
            MDC.put("CORRELATION_ID", enrichedKafkaEvent.getCorrelationId());
            functionalReqFailCounter.labels(reason, enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getKafkaEvent().getMottaksKanal(), String.valueOf(enrichedKafkaEvent.getSkjema())).inc();
        }catch (Exception e){
            log.error("Klarte ikke å inkrementere metrikk incFailFunctionalRequirements med verdier: '{}' '{}' '{}' '{}'", reason, enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getKafkaEvent().getMottaksKanal(), enrichedKafkaEvent.getSkjema(), e);
        }
    }

    public static void incInvalidJournalpostStatus(EnrichedKafkaEvent enrichedKafkaEvent){
        try {
            MDC.put("CORRELATION_ID", enrichedKafkaEvent.getCorrelationId());
            invalidJournalpostStatusCounter.labels(enrichedKafkaEvent.getJournalpost().getJournalstatus(), enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getKafkaEvent().getMottaksKanal(), String.valueOf(enrichedKafkaEvent.getSkjema())).inc();
        }catch (Exception e){
            log.error("Klarte ikke å inkrementere metrikk incInvalidJournalpostStatus med verdier: '{}' '{}' '{}' '{}'", enrichedKafkaEvent.getJournalpost().getJournalstatus(), enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getKafkaEvent().getMottaksKanal(), enrichedKafkaEvent.getSkjema(), e);
        }
    }

    public static void incRetry(String transformer, EnrichedKafkaEvent enrichedKafkaEvent){
        String skjema = "null";
        try {
            MDC.put("CORRELATION_ID", enrichedKafkaEvent.getCorrelationId());
            if(enrichedKafkaEvent.getJournalpost() != null ){
                skjema = enrichedKafkaEvent.getSkjema();
            }
            retryCounter.labels(transformer, String.valueOf(enrichedKafkaEvent.getNumFailedAttempts()), enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getKafkaEvent().getMottaksKanal(), String.valueOf(enrichedKafkaEvent.getSkjema())).inc();
        }catch (Exception e){
            log.error("Klarte ikke å inkrementere metrikk retryCounter med verdier: '{}' '{}' '{}' '{}' '{}'", transformer, String.valueOf(enrichedKafkaEvent.getNumFailedAttempts()), enrichedKafkaEvent.getTema(), enrichedKafkaEvent.getKafkaEvent().getMottaksKanal(), skjema, e);
        }
    }

    public static void setRetrystoreGauge(String name, long num){
        try {
            retrystoreGauge.labels(name).set(num);
        }catch (Exception e){
            log.error("Klarte ikke sette retry gauge");
        }
    }
}

File: src/main/kotlin/no/nav/helse/flex/infrastructure/metrics/Metrics.kt
package no.nav.helse.flex.infrastructure.metrics

import io.prometheus.client.Counter
import io.prometheus.client.Gauge
import no.nav.helse.flex.infrastructure.kafka.EnrichedKafkaEvent
import org.slf4j.LoggerFactory
import org.slf4j.MDC

object Metrics {
    private val log = LoggerFactory.getLogger(Metrics::class.java)
    private const val MANUELL = "Manuell journalforing"
    private const val AUTO = "Automatisk journalforing"

    private val jfrProcessCounter = Counter.Builder()
        .name("flex_joark_mottak_process_counter")
        .help("Teller antall journalposter som behandles. Setter på labels på om det burde blitt behandlet maskinelt/manuelt og om den blir behandlet maskinelt/manuelt")
        .labelNames("result", "desired", "tema", "kanal", "skjema")
        .register()

    private val feilregOppgaverCounter = Counter.Builder()
        .name("flex_joark_mottak_feilregistrerte_oppgaver_counter")
        .help("Teller antall ganger vi må avbryte autoatiske prosess på en slik måte at en oppgave må feilregistreres")
        .labelNames("ableToFeilreg", "tema", "kanal", "skjema")
        .register()

    private val functionalReqFailCounter = Counter.Builder()
        .name("flex_joark_mottak_functional_req_fail_counter")
        .help("Teller funksjonelle hindringer for automatisk journalforing")
        .labelNames("reason", "tema", "kanal", "skjema")
        .register()

    private val invalidJournalpostStatusCounter = Counter.Builder()
        .name("flex_joark_mottak_invalid_journalpost_status_counter")
        .help("Teller journalposter fra SAF med ugyldig status")
        .labelNames("status", "tema", "kanal", "skjema")
        .register()

    private val retryCounter = Counter.Builder()
        .name("flex_joark_mottak_retry_counter")
        .help("Teller antall journalposter hver gang vi vil prøve på nytt senere")
        .labelNames("transformer", "numErrors", "tema", "kanal", "skjema")
        .register()

    private val retrystoreGauge = Gauge.Builder()
        .name("flex_joark_mottak_retry_gauge")
        .help("Viser hvor mange JP som ligger på retrystore")
        .labelNames("transformer")
        .register()

    private val jfrManuellProcessCounter = Counter.build(
        "jfr_manuell_process_counter",
        "Teller antall manuelle journalposter som behandles."
    )
        .labelNames("result", "desired", "tema")
        .register()

    private val createOppgaveCounter = Counter.build(
        "jfr_manuell_create_oppgaver_counter",
        "Teller antall manuelle oppgaver som opprettes. Diffrensierer på fordeling og journalføringsoppgave"
    )
        .labelNames("oppgavetype", "tema")
        .register()

    fun incJfrManuallProcess(enrichedKafkaEvent: EnrichedKafkaEvent, desiredAutomaticJfr: Boolean) {
        val desired: String
        var skjema = "null"

        try {
            MDC.put("CORRELATION_ID", enrichedKafkaEvent.correlationId)

            desired = if (desiredAutomaticJfr) AUTO else MANUELL

            if (enrichedKafkaEvent.journalpost != null && enrichedKafkaEvent.skjema != null) {
                skjema = enrichedKafkaEvent.skjema!!
            }

            jfrProcessCounter.labels(
                MANUELL,
                desired,
                enrichedKafkaEvent.tema,
                enrichedKafkaEvent.kafkaEvent.mottaksKanal,
                skjema
            ).inc()
        } catch (e: Exception) {
            log.error("Klarte ikke å inkrementere metrikk incJfrManuallProcess", e)
        }
    }

    fun incJfrAutoProcess(enrichedKafkaEvent: EnrichedKafkaEvent) {
        try {
            MDC.put("CORRELATION_ID", enrichedKafkaEvent.correlationId)
            jfrProcessCounter.labels(
                AUTO,
                AUTO,
                enrichedKafkaEvent.tema,
                enrichedKafkaEvent.kafkaEvent.mottaksKanal,
                enrichedKafkaEvent.skjema.toString()
            ).inc()
        } catch (e: Exception) {
            log.error("Klarte ikke å inkrementere metrikk incJfrAutoProcess", e)
        }
    }

    fun incFeilregCounter(enrichedKafkaEvent: EnrichedKafkaEvent, sucessfullFeilreg: Boolean) {
        try {
            MDC.put("CORRELATION_ID", enrichedKafkaEvent.correlationId)
            feilregOppgaverCounter.labels(
                java.lang.Boolean.toString(sucessfullFeilreg),
                enrichedKafkaEvent.tema,
                enrichedKafkaEvent.kafkaEvent.mottaksKanal,
                enrichedKafkaEvent.skjema.toString()
            ).inc()
        } catch (e: Exception) {
            log.error("Klarte ikke å inkrementere metrikk incFeilregCounter", e)
        }
    }

    fun incFailFunctionalRequirements(reason: String?, enrichedKafkaEvent: EnrichedKafkaEvent) {
        try {
            MDC.put("CORRELATION_ID", enrichedKafkaEvent.correlationId)
            functionalReqFailCounter.labels(
                reason,
                enrichedKafkaEvent.tema,
                enrichedKafkaEvent.kafkaEvent.mottaksKanal,
                enrichedKafkaEvent.skjema.toString()
            ).inc()
        } catch (e: Exception) {
            log.error("Klarte ikke å inkrementere metrikk incFailFunctionalRequirements", e)
        }
    }

    fun incInvalidJournalpostStatus(enrichedKafkaEvent: EnrichedKafkaEvent) {
        try {
            MDC.put("CORRELATION_ID", enrichedKafkaEvent.correlationId)
            invalidJournalpostStatusCounter.labels(
                enrichedKafkaEvent.journalpost?.journalstatus,
                enrichedKafkaEvent.tema,
                enrichedKafkaEvent.kafkaEvent.mottaksKanal,
                enrichedKafkaEvent.skjema.toString()
            ).inc()
        } catch (e: Exception) {
            log.error("Klarte ikke å inkrementere metrikk incInvalidJournalpostStatus", e)
        }
    }

    fun incRetry(transformer: String?, enrichedKafkaEvent: EnrichedKafkaEvent) {
        try {
            MDC.put("CORRELATION_ID", enrichedKafkaEvent.correlationId)
            retryCounter.labels(
                transformer,
                enrichedKafkaEvent.numFailedAttempts.toString(),
                enrichedKafkaEvent.tema,
                enrichedKafkaEvent.kafkaEvent.mottaksKanal,
                enrichedKafkaEvent.skjema.toString()
            ).inc()
        } catch (e: Exception) {
            log.error("Klarte ikke å inkrementere metrikk incRetry", e)
        }
    }

    fun setRetrystoreGauge(name: String?, num: Long) {
        try {
            retrystoreGauge.labels(name).set(num.toDouble())
        } catch (e: Exception) {
            log.error("Klarte ikke sette retry gauge")
        }
    }

    fun incProcessCounter(result: String?, desireResult: String?, tema: String?) {
        jfrManuellProcessCounter.labels(result, desireResult, tema).inc()
    }

    fun incCreateOppgaveCounter(oppgavetype: String?, tema: String?) {
        createOppgaveCounter.labels(oppgavetype, tema).inc()
    }
}
